LibreNMS/Modules/Links.php

```diff
@@ -0,0 +1,317 @@
+<?php
+
+namespace LibreNMS\Modules;
+
+use App\Facades\LibrenmsConfig;
+use App\Facades\PortCache;
+use App\Models\Device;
+use App\Models\Link;
+use App\Models\Port;
+use App\Observers\ModuleModelObserver;
+use Illuminate\Support\Collection;
+use LibreNMS\DB\SyncsModels;
+use LibreNMS\Interfaces\Data\DataStorageInterface;
+use LibreNMS\Interfaces\Discovery\LinkDiscovery;
+use LibreNMS\Interfaces\Module;
+use LibreNMS\OS;
+use LibreNMS\Polling\ModuleStatus;
+use LibreNMS\Util\IP;
+use LibreNMS\Util\Mac;
+use LibreNMS\Util\StringHelpers;
+use SnmpQuery;
+
+class Links implements Module
+{
+    use SyncsModels;
+
+    /**
+     * @inheritDoc
+     */
+    public function dependencies(): array
+    {
+        return ['ports'];
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function shouldDiscover(OS $os, ModuleStatus $status): bool
+    {
+        return $status->isEnabledAndDeviceUp($os->getDevice());
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function shouldPoll(OS $os, ModuleStatus $status): bool
+    {
+        return false;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function discover(OS $os): void
+    {
+        $links = new Collection;
+        $group = $os->getDeviceArray()['os_group'] ?? '';
+
+        if ($os instanceof LinkDiscovery) {
+            $links = $os->discoverLinks();
+        }
+
+        if ($links->isEmpty() || $group == 'cisco') {
+            $links = $links->merge($this->discoverLldp($os->getDevice()));
+        }
+
+        $links->each(function (Link $link) use ($os) {
+            $link->remote_hostname = substr((string) $link->remote_hostname, 0, 127);
+            $link->remote_port = substr((string) $link->remote_port, 0, 127);
+            $link->remote_version = substr((string) $link->remote_version, 0, 255);
+            $link->remote_platform = substr((string) $link->remote_platform, 0, 255);
+
+            $tmp = explode('#', $link->protocol);
+            $link->protocol = $dp = $tmp[0];
+            $ip = $tmp[1] ?? '';
+
+            if (empty($link->remote_device_id) &&
+                \LibreNMS\Util\Validate::hostname($link->remote_hostname) &&
+                ! can_skip_discovery($link->remote_hostname, $link->remote_version) &&
+                LibrenmsConfig::get('autodiscovery.xdp') === true) {
+                $link->remote_device_id = discover_new_device(
+                    $link->remote_hostname, $os->getDeviceArray(), strtoupper($dp), Port::where('port_id', $link->local_port_id)->first()->toArray()
+                ) ?: 0;
+            }
+            if (empty($link->remote_device_id) &&
+                ! empty($ip) &&
+                LibrenmsConfig::get('discovery_by_ip', false) &&
+                LibrenmsConfig::get('autodiscovery.xdp') === true) { //name lookup failed, try with IP
+                $link->remote_device_id = discover_new_device(
```

Comment:
Can this code even call these legacy functions ?

Reply:
hi @Jellyfrog 

yes, it is working, tested many times

Reply:
To add a little extra. I greatly appreciate contributions.   If you want things merged faster, make them easier to merge.